lusidtools/cocoon/cocoon.py

```diff
@@ -901,11 +904,342 @@ async def _construct_batches(
+        )
+
+        next_page = response.next_page
+        done = response.next_page is None
+
+    return unmatched_transactions
+
+
+def filter_unmatched_transactions(
+    data_frame: pd.DataFrame, mapping_required: dict, unmatched_transactions: list,
+):
+    """
+    This method will take the full list of unmatched transactions and remove any transactions that were not
+    part of the current upload with load_from_data_frame.
+
+    Parameters
+    ----------
+    data_frame : pd.DataFrame
+        The DataFrame containing the data
+    mapping_required : dict
+        The required mapping from load_from_data_frame that helps identify the transaction_id column in the DataFrame
+    unmatched_transactions : list
+         A list of objects with the structure { transaction_id: instrument_identifiers }
+
+    Returns
+    -------
+    A filtered list of unmatched_transactions.
+    """
+    # Create a list var to hold the filtered transactions to be returned
+    filtered_unmatched_transactions = []
+
+    # Create a unique list of transaction_ids from the dataframe
+    valid_txids = data_frame[mapping_required.get("transaction_id")].unique()
+
+    # Iterate through the transactions and if the transaction_id is in the dataframe, add it to the list to be returned
+    for unmatched_transaction in unmatched_transactions:
+        for key in unmatched_transaction.keys():
+            if key in valid_txids:
+                filtered_unmatched_transactions.append(unmatched_transaction)
+
+    return filtered_unmatched_transactions
+
+
+def _unmatched_ids_holdings(
+    api_factory: lusid.utilities.ApiClientFactory,
+    scope: str,
+    sync_batches: list = None,
+):
+    """
+    This method identifies which instruments were not resolved with a holdings upload using load_from_data_frame.
+
+    Parameters
+    ----------
+    api_factory : lusid.utilities.ApiClientFactory api_factory
+        The api factory to use
+    scope : str
+        The scope of the resource to load the data into
+    sync_batches : list
+        A list of the batches used to upload the data into LUSID
+
+    Returns
+    -------
+    responses: list
+        A list to be appended to the ultimate response for load_from_data_frame.
+
+        The returned list will be a unique list of instrument_identifiers.
+
+    """
+    # Extract a list of tuples of portfolio codes and effective at times from sync_batches
+    code_tuples = _extract_unique_portfolio_codes_effective_at_tuples(sync_batches)
+
+    # Create empty list to hold instrument identifiers that have not been resolved
+    unmatched_instruments = []
+
+    # For each holding adjustment in the upload, check whether any contained unresolved instruments and append to list
+    for code_tuple in code_tuples:
+        unmatched_instruments.extend(
+            return_unmatched_instruments_from_holdings(
+                api_factory=api_factory, scope=scope, code_tuple=code_tuple,
+            )
+        )
```

Comment:
Same point of using a list comprehension as I made above

Reply:
Similar to previous comments, return_unmatched_instruments_from_holdings returns a list of objects. If we use a list comprehension we would end up with a list of lists, which would need flattening. This current logic is not more complex than that.